Applications/Loader/main.cpp

```diff
@@ -0,0 +1,252 @@
+#include <LibELF/exec_elf.h>
+#include <dlfcn.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+// #define DYNAMIC_LOAD_VERBOSE
+
+#ifdef DYNAMIC_LOAD_VERBOSE
+#    define VERBOSE(fmt, ...) dbgprintf(fmt, ##__VA_ARGS__)
+#else
+#    define VERBOSE(fmt, ...) \
+        do {                  \
+        } while (0)
+#endif
+#define TLS_VERBOSE(fmt, ...) dbgprintf(fmt, ##__VA_ARGS__)
+
+char* __static_environ[] = { nullptr }; // We don't get the environment without some libc workarounds..
+
+static HashMap<String, NonnullRefPtr<ELF::DynamicLoader>> g_loaders;
+static HashMap<String, NonnullRefPtr<ELF::DynamicObject>> g_loaded_objects;
+
+static size_t g_current_tls_offset = 0;
+static size_t g_total_tls_size = 0;
+
+static void init_libc()
+{
+    environ = __static_environ;
+    __environ_is_malloced = false;
+    __stdio_is_initialized = false;
+    __malloc_init();
+}
+
+static void perform_self_relocations()
+{
+    // We need to relocate ourselves.
+    // (these relocations seem to be generated because of our vtables)
+
+    // TODO: Pass this address in the Auxiliary Vector
+    u32 base = 0x08000000;
+    Elf32_Ehdr* header = (Elf32_Ehdr*)(base);
+    Elf32_Phdr* pheader = (Elf32_Phdr*)(base + header->e_phoff);
+    u32 dynamic_section_addr = 0;
+    for (size_t i = 0; i < (size_t)header->e_phnum; ++i, ++pheader) {
+        if (pheader->p_type != PT_DYNAMIC)
+            continue;
+        dynamic_section_addr = pheader->p_vaddr + base;
+    }
+    if (!dynamic_section_addr)
+        exit(1);
+
+    auto dynamic_object = ELF::DynamicObject::construct((VirtualAddress(base)), (VirtualAddress(dynamic_section_addr)));
+
+    dynamic_object->relocation_section().for_each_relocation([base](auto& reloc) {
+        if (reloc.type() != R_386_RELATIVE)
+            return IterationDecision::Continue;
+        *(u32*)reloc.address().as_ptr() += base;
+        return IterationDecision::Continue;
+    });
+}
+
+static ELF::DynamicObject::SymbolLookupResult global_symbol_lookup(const char* symbol_name)
+{
+    VERBOSE("global symbol lookup: %s\n", symbol_name);
+    for (auto& lib : g_loaded_objects) {
+        VERBOSE("looking up in object: %s\n", lib.key.characters());
+        auto res = lib.value->lookup_symbol(symbol_name);
+        if (!res.has_value())
+            continue;
+        return res.value();
+    }
+    ASSERT_NOT_REACHED();
+    return {};
+}
+
+static void map_library(const String& name, int fd)
+{
+    struct stat lib_stat;
+    int rc = fstat(fd, &lib_stat);
+    ASSERT(!rc);
+
+    auto loader = ELF::DynamicLoader::construct(name.characters(), fd, lib_stat.st_size);
+    loader->set_tls_offset(g_current_tls_offset);
+    loader->m_global_symbol_lookup_func = global_symbol_lookup;
+
+    g_loaders.set(name, loader);
+
+    g_current_tls_offset += loader->tls_size();
+}
+
+static void map_library(const String& name)
+{
+    // TODO: Do we want to also look for libs in other paths too?
```

Comment:
Room here for LD_LIBRARY_PATH, RUNPATH, RPATH, ... But definitely not first pass.